### PR TITLE
Implement memory registration framework

### LINKED ISSUE #147: Add basic memory registration (fi_mr_ops)
mr_reg and mr_dereg needs to be implemented for basic send/recv operations

@hppritcha, @ztiffany , @sungeunchoi 

Please add more content, suggestions, etc ...

### PR DESCRIPTION
**Pull request by jswaro**
This commit implements the memory registration work for the GNIX provider
at a very basic level. Additional work must be done to support different
progress models and eventually a registration cache.

Signed-off-by: James Swaro jswaro@cray.com

```
modified:   prov/gni/Makefile.am
new file:   prov/gni/include/gnix_mr.h
modified:   prov/gni/src/gnix_dom.c
modified:   prov/gni/src/gnix_mr.c
new file:   prov/gni/test/mr.c
```

closes #147 

@hppritcha , @ztiffany

### CODE REVIEW

**Comment by jswaro** [comment]
The PR is still in an early state but I'd like to get the comments and reviews started sooner rather than later. Let me know what you think may still need changing.

File: prov/gni/include/gnix_mr.h [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+					uint64_t va: GNIX_MR_VA_BITS;
+					uint64_t mdd: GNIX_MR_MDD_BITS;
+				};
+			};
+			uint64_t flags : GNIX_MR_FLAG_BITS;
+			uint64_t padding: GNIX_MR_PADDING_LENGTH;
+		};
+		uint64_t value;
+	};
+} gnix_mr_key_t;
```
Comment (ztiffany) [inline prov/gni/include/gnix_mr.h:76]: I think we should rename 'va' to 'pfn' in this structure.  Do you need to name the field of va and mdd combined?

File: prov/gni/src/gnix_mr.c [hunk 2]
```diff
@@ -38,11 +38,210 @@
...
+	atomic_inc(&mr->ref_cnt);
+}
+
+__attribute__((unused))
+static void gnix_convert_key_to_mhdl(gnix_mr_key_t *key, gni_mem_handle_t *mhdl)
+{
+	/*
+
+	uint64_t va = (uint64_t) sign_extend(key->va << PAGE_SHIFT,
+			GNIX_MR_KEY_BITS);
```
Comment (ztiffany) [inline prov/gni/src/gnix_mr.c]: You want the size of the va field shifted PAGE_SIZE here, not GNIX_MR_KEY_BITS.

**Comment by hppritcha** [comment]
This looks like a good start.  Do we need to worry about the old MDH format?

File: prov/gni/src/gnix_mr.c [hunk 3]
```diff
@@ -38,11 +38,210 @@
...
+	{
+
+		/* TODO: when we determine the progress model,
+		 *         set the cq_hndl pointer
+		 */
+		grc = GNI_MemRegister(nic->gni_nic_hndl, (uintptr_t) buf, len,
+					cq_hndl, fi_gnix_access,
+					-1, &mr->md.mem_hndl);
+		if (grc == GNI_RC_SUCCESS)
+			break;
```
Comment (hppritcha) [inline prov/gni/src/gnix_mr.c:154]: i'm pretty sure we're not going to need to use a cq_hndl for memory registered using the fi_mr_reg interface.

Comment (jswaro) [inline prov/gni/src/gnix_mr.c:154]: I can just replace that with a NULL value instead.
-- Jim

**Comment by jswaro** [comment]
I'm not sure. @ztiffany

-- Jim

**Comment by jswaro** [comment]
@hppritcha , @ztiffany 

Should we move the gni_priv macros and structures to a separate file?

File: prov/gni/src/gnix_mr.c [hunk 4]
```diff
@@ -38,11 +38,372 @@
...
+
+		/* TODO: when we determine the progress model,
+		 *         set the cq_hndl pointer
+		 */
+		grc = GNI_MemRegister(nic->gni_nic_hndl, (uintptr_t) buf, len,
+					NULL, fi_gnix_access,
+					-1, &mr->mem_hndl);
+		if (grc == GNI_RC_SUCCESS)
+			break;
+	}
```
Comment (hppritcha) [inline prov/gni/src/gnix_mr.c:155]: need to add code to allocate a nic if there aren't any yet on the domain's nic list.

**Comment by sungeunchoi** [comment]
retest this please

**Comment by hppritcha** [comment]
Refer to this link for build results (access rights to CI server needed): 
jenkins.open-mpi.org/job/edison_libfabric_pr/44/
Test PASSed.

**Comment by hppritcha** [comment]
yeah maybe putting those special macros in a different file would be good.  better shows a dependency on something outside of gni public api right now.

**Comment by hppritcha** [comment]
Refer to this link for build results (access rights to CI server needed): 
jenkins.open-mpi.org/job/Aggregator%20Bot%20Test/1/
Test PASSed.

File: prov/gni/src/gnix_mr.c (hunk 4 above)
Comment (jswaro) [inline prov/gni/src/gnix_mr.c:155]: Done. See line 138

File: prov/gni/include/gnix_mr.h [hunk 5]
```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) 2015 Cray Inc. All rights reserved.
+ *
```
Comment (sungeunchoi) [inline prov/gni/include/gnix_mr.h:3]: @hppritcha Do you know if the license must be included in every file?  I don't see that on the libfabric wiki, but regardless, it seems like a good idea.

**Comment by hppritcha** [comment]
retest this please

**Comment by hppritcha** [comment]
i was just checking to see that the IU jenkins was still working

**Comment by hppritcha** [comment]
Refer to this link for build results (access rights to CI server needed): 
jenkins.open-mpi.org/job/edison_libfabric_pr/47/
Test PASSed.

File: prov/gni/include/gnix_mr.h [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+#define GNIX_MR_H_
+
+#include "gnix.h"
+#include "gnix_nic.h"
+#include "gnix_util.h"
+#include "ccan/list.h"
+
+#define GNIX_MR_PAGE_SHIFT 12
+#define GNIX_MR_PFN_BITS 37
+#define GNIX_MR_MDD_BITS GNIX_MR_PAGE_SHIFT
```
Comment (ztiffany) [inline prov/gni/include/gnix_mr.h]: The 12 bit MDD is not related to the 4k page size.  I don't think you should define MDD_BITS with PAGE_SHIFT.

File: prov/gni/src/gnix_mr.c [hunk 7]
```diff
@@ -38,11 +38,180 @@
...
+
+	return r;
+}
+
+void gnix_convert_key_to_mhdl(
+		IN    gnix_mr_key_t *key,
+		INOUT gni_mem_handle_t *mhdl)
+{
+	uint64_t pfn = (uint64_t) __sign_extend(key->pfn << PAGE_SHIFT,
+			GNIX_MR_PFN_BITS);
```
Comment (ztiffany) [inline prov/gni/src/gnix_mr.c]: You should sign extend from PFN_BITS+PAGE_SHIFT, or 49 bits.  The result is a virtual address, not pfn.

File: prov/gni/src/gnix_mr.c [hunk 8]
```diff
@@ -38,11 +38,180 @@
...
+	GNI_MEMHNDL_SET_FLAGS((*mhdl), flags);
+	GNI_MEMHNDL_SET_PAGESIZE((*mhdl), PAGE_SHIFT);
+	GNI_MEMHNDL_SET_CRC((*mhdl));
+}
+
+void gnix_convert_mhdl_to_key(
+		IN    gni_mem_handle_t *mhdl,
+		INOUT gnix_mr_key_t *key)
+{
+	key->pfn = GNI_MEMHNDL_GET_VA((*mhdl));
```
Comment (ztiffany) [inline prov/gni/src/gnix_mr.c]: The 37 bit pfn should be the mem handle VA right shifted by PAGE_SHIFT.

**Comment by hppritcha** [comment]
looks good.  I opened issue #165 to track getting the atomics back upstream.

**Comment by hppritcha** [comment]
another bot test

bot:retest

**Comment by hppritcha** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.open-mpi.org/job/edison_libfabric_pr/48/
Test PASSed.

**Comment by jswaro** [comment]
So where do we stand on this PR? 

@hppritcha , @ztiffany , @sungeunchoi

**Comment by hppritcha** [comment]
I think this is ready for merging.

**Pull request merged**